### DISCUSSION TITLE
specific heat for mixtures

### DISCUSSION THREAD

**Discussion by alexDfluid** [usnistgov/REFPROP-issues#578] [Q&A]
Hello guys,

I would like to know if it is possible to calculate the specific heats c_v and c_p for mixtures in refprop using the python interface?
For example:
Propane - Nitrogen mixture
mixture ratio z [0.5,0.5]
temperature T 
pressure P

I use TPFLSH which gives reasonable results for the single phase region.
But the documentation mentions, that cp is not defined for 2-phase states and Cv for 2-phase states is not calculated.
As far as I've seen Cv2phase is not applicable to mixtures, but only for pure substances?

Any possibility to get the specific heats for mixtures?

best,
Alex

**Comment by ianhbell** [comment]
c_p is not defined for two-phase states for pure fluids. Thinking a little bit about your question, I can agree that one *should* be able to define a c_p for a two-phase mixture because dh/dT|P is not ill-defined for a two-phase VLE mixture. But REFPROP does not currently support such a calculation. How you *could* do this is to take your inputs from REFPROP and plug them into CoolProp, which has the ``first_partial_deriv`` function that should in theory give you what you are asking for (I think).

**Reply by ianhbell** [reply]
Now that I say that, there is also this function: https://refprop-docs.readthedocs.io/en/latest/DLL/legacy.html#f/_/DHD1dll

**Reply by ianhbell** [reply]
So you would calculate the density as usual, and then plug that density into DHD1dll function

**Reply by alexDfluid** [reply]
Thank you for the very fast reply.
I tried to use DHD1dll in the two phase region where 0 <= q <=1 for the Propane/Nitrogen system at 280K and 50bar.
Unfortunately, there seem to be a discontinuity.
![grafik](https://user-images.githubusercontent.com/111732828/216317113-a2e77fe3-fcc2-485b-8395-fe1916c552a5.png)

Any further suggestions how to get c_v and c_p in a two phase mixture?

**Reply by ianhbell** [reply]
Hmm, I am not entirely sure the very *idea* of a two-phase c_p is well-posed for a mixture.

**Comment by nist-aharvey** [comment]
Cp and Cv are both defined for a mixture in the 2-phase region as just their normal definitions, taken with constant overall composition. So for example Cp is (dh/dT)_p. [I think 2-phase Cp may become undefined if you are at an azeotrope, but otherwise should be a well-defined quantity.]
However, I don't think there is any straightforward way to calculate it; because vaporization and composition changes in the phases happen when heat is added it is not anything simple like a combination of (dh/dT)_p of the individual phases.
So my only suggestion would be to evaluate the derivative as a finite difference, being careful that you keep the appropriate variables constant when you increment the temperature.
To be safe, I would also apply the flag to set the reference state to "each pure component" since sometimes mixture reference states can cause unexpected results when compositions are changing.

**Reply by alexDfluid** [reply]
Thank you for your response.
I was hoping for a more simple solution for the calculation of the specific heats, but it seems that I will have to apply your suggestion.

**Comment by dnwillia-work** [comment]
I just ran into this same issue this week. Say you have a two phase state then the liquid and gas cp values for an ideal solution are well defined:

$$c_{pL} = \sum_{n=1}^{N_c}x_n \ c_{pL,n}$$
$$c_{pV} = \sum_{n=1}^{N_c}y_n \ c_{pV,n}$$

and if you are at a particular $(T, P, z_i)$ in the 2-phase region, and do a flash iteration to find the quality then enthalpy is

$$h_{mix} = q \ h_V + (1 - q) \ h_L$$

if you take the partial derivative w.r.t. temperature at constant pressure

$$c_{pmix} = q \ c_{pV} + (1 - q) \ c_{pL} + (h_V-h_L)\left(\frac{\partial q}{\partial T}\right)_{p, z_i}$$

So, for an ideal solution it's certainly not a simple matter of weighting the liquid and gas specific heat by the quality.  It depends on the latent heat and $q(T)$.  REFPROP can compute the two phase flash and output the quality, and one can verify it's a function of temperature.

I'd say for ideal solutions at least, it's tractable to compute it, but for now, even for that, the direction we went for the last term is to estimate it numerically.

I'm not sure what this looks like for a real mixture, certainly much more complicated perhaps, but I would be surprised if there is no definition for a 2 phase mixture at all as mentioned by @ianhbell.  The thing is that even for a real mixture cp is perfectly well defined right up to the dew and bubble points, so why not in between?  Mathematically speaking an $h(T, P_{const}, z_i)$ curve for a real mixture will certainly have a slope in the two phase region that is the specific heat at constant pressure.

**Reply by nist-aharvey** [reply]
While I don't have time to look into it in detail, I think this approach is not quite correct (whether the mixture is ideal or not does not really matter here).
It does not account for the change in composition of the vapor and liquid as you move to a different temperature, which makes the changes in hv and hL with temperature different from simply the cp of the saturated phases.
If you are estimating a term numerically anyway, you might as well do the whole thing numerically based on the definition of cp, by doing something like computing h(T,p) and h(T+0.001,p) where h is the enthalpy of the entire 2-phase system.
